packages/@okta/vuepress-site/docs/guides/authenticators-google-authenticator/main/android/enrollmentintegrationsteps.md

```diff
@@ -0,0 +1,447 @@
+    // If no authenticators are found for challenge show error and abort
+    if (remediation.authenticators.size == 0) {
+        // handle error
+        return
+    }
+    val authenticator = remediation.authenticators.first()
+
+    when (authenticator.type) {
+        IdxAuthenticator.Kind.PASSWORD -> {
+            // Update the value in the remediation object and proceed to the next step in IDX flow
+            remediation["credentials.passcode"]?.value = password
+            remediation.proceed()
+        }
+        // handle other authenticators
+        else -> {
+            // handle unknown authenticator error
+            return
+        }
+    }
+}
+```
+
+> **Note:** The response from the server specifies either the next step, an error, or a
+> message, such as an incorrect code. The code that handles the response may need to show
+> different UI, depending on the step. In the example app [`LoginActivity`](https://github.com/okta-samples/okta-android-oie-authenticators-quickstart/blob/main/app/src/main/java/com/okta/android/samples/authenticator/ui/login/LoginActivity.kt) handles receiving a sign-in token in case of success, receiving errors and rendering dynamic UI fields.
+
+### 4: Display a list of possible authenticator factors
+
+After verifying the username and password, the server sends a remediation of type `SELECT_AUTHENTICATOR_ENROLL` that contains the possible authenticators. The list of choices are in the `options` property of a field with `name` set to `"authenticator"` in the remediation's `forms`, `visibleFields` property.
+
+The display name of an authenticator is in the `label` property of the field. Use this name to build your selection UI. See the `IdxRemediation.Form.Field.asIdxDynamicFields()` function in the [LoginViewModel class of the sample application](https://github.com/okta-samples/okta-android-oie-authenticators-quickstart/blob/main/app/src/main/java/com/okta/android/samples/authenticator/ui/login/LoginViewModel.kt) for an example of building a dynamic field from options.
+
+When the user selects an option, set the `selectedOption` property of the field containing the options (the one with the `name` `"authenticator"`) to the field containing the selected option. Call `proceed()` on the remediation once the user has finalized their choice to send the information and request the next step.
+
+You may receive multiple enrollment requests during the sign-on flow when a policy includes more than one authenticator. When the policy allows the user to skip enrolling in additional authenticators the server returns at least one remediation with a `type` of `IdxRemediation.Type.SKIP` and you can use that to display a skip action and call proceed on that remediation.
+
+```kotlin
+private suspend fun handleAuthenticatorEnrollOrChallenge(remediation: IdxRemediation, skipRemediation: IdxRemediation?) {
+    val fields = mutableListOf<IdxDynamicField>()
```

Comment:
There's a lot of usage of `IdxDynamicField` here, which isn't part of the SDK, and is never defined. As a first time reader, this might be confusing.

Reply:
I thought I explained it but seems I missed i'll add an explanation before first use. Btw this could be an useful data structure we could provide in the SDK right?

Reply:
Something similar might be useful in another layer of SDK, but not the current one.

For instance, we hope to build a sign in widget for native mobile. Which would likely include a "middle layer" SDK that would include something with a similar layer of abstraction.